k8sclient/k8sclient.go

```diff
@@ -69,6 +79,54 @@ func createK8sClient(kubeconfig string) (KubeClient, error) {
 	return &defaultKubeClient{client: client}, nil
 }
 
+func setKubeClientInfo(c *ClientInfo, client KubeClient, k8sArgs types.K8sArgs) {
+	c.Client = client
+	c.Podnamespace = string(k8sArgs.K8S_POD_NAMESPACE)
+	c.Podname = string(k8sArgs.K8S_POD_NAME)
+}
+
+func SetNetworkStatus(k *ClientInfo, netStatus []*types.NetworkStatus) error {
+
+	pod, err := k.Client.GetPod(k.Podnamespace, k.Podname)
+	if err != nil {
+		return fmt.Errorf("SetNetworkStatus: failed to query the pod %v in out of cluster comm: %v", k.Podname, err)
+	}
+
+	var networkStatus []string
+	for _, nets := range netStatus {
+		data, err := json.MarshalIndent(nets, "", "    ")
+		if err != nil {
+			return fmt.Errorf("SetNetworkStatus: error with Marshal Indent: %v", err)
+		}
+		networkStatus = append(networkStatus, string(data))
+	}
+
+	ns := fmt.Sprintf("[%s]", strings.Join(networkStatus, ","))
+
+	_, err = setPodNetworkAnnotation(k.Client, k.Podnamespace, pod, ns)
+	if err != nil {
+		return fmt.Errorf("SetNetworkStatus: failed to update the pod %v in out of cluster comm: %v", k.Podname, err)
+	}
+
+	return nil
+}
+
+func setPodNetworkAnnotation(client KubeClient, namespace string, pod *v1.Pod, networkstatus string) (*v1.Pod, error) {
+	//if pod annotations is empty, make sure it allocatable
+	if len(pod.Annotations) == 0 {
+		pod.Annotations = make(map[string]string)
+	}
+
+	pod.Annotations["kubernetes.v1.cni.cncf.io/networks-status"] = networkstatus
+
+	p, err := client.SetPod(namespace, pod)
```

Comment:
Instead of doing this, I would use the Kube util ExponentialBackoff() to try updating a couple times.  Just doing it once is somewhat error-prone in that if anything else updates the Pod object your update will get rejected.  Instead we can use RetryOnConflict:

```
import (
	"k8s.io/client-go/util/retry"
)


func (d *defaultKubeClient) UpdatePodStatus(pod *v1.Pod) (*v1.Pod, error) {
	return d.client.Core().Pods(pod.Namespace).UpdateStatus(pod)
}


	pod := pod.DeepCopy()
	var err error
	if resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
		if err != nil {
			// Re-get the pod unless it's the first attempt to update
			pod, err = client.GetPod(namespace, name)
			if err != nil {
				return err
			}
		}

		pod, err = client.UpdatePodStatus(pod)
		return err
	}); resultErr != nil {
		return nil, fmt.Errorf("status update failed for pod %s/%s: %v", namespace, name, resultErr)
	}
	return pod, nil
```

Reply:
@dcbw I was exactly looking for some solution like that. I will try this.